### ISSUE TITLE
JPEG 7 compression - wrong colors

### ISSUE THREAD

**Issue by photopea** [photopea/UTIF.js#42]
Following TIFFs with Type7 compression have incorrect colors. The PhotometricInterpretation is 2 (RGB), however, colors seem to be more like L*a*b*.

[tiff.zip](https://github.com/photopea/UTIF.js/files/2320463/tiff.zip)

**Comment by photopea** [comment]
@syvaidya  could you take a look at it? Maybe the info about colors is inside the JPEG block.

**Comment by syvaidya** [comment]
I'll try to check tomorrow. Been really busy past few days.

**Comment by syvaidya** [comment]
@photopea sorry for being so late, but finally i am on vacation and got time to check this. The problem seems to be with JpegDecoder. If I don't decode the "buff" and directly save it as "jpg" files, then that image is shown with correct colors in any image viewer. It looks like that JpegDecoder is messing up the colors.

**Comment by mahmoodels** [comment]
Hi,
I have the same issue in the following file [sample.zip](https://github.com/photopea/UTIF.js/files/3426261/sample.zip), so how can i fix this issue?

**Comment by skibulk** [comment]
I am also having this issue. Using the file [here](https://www.nightprogrammer.org/development/multipage-tiff-example-download-test-image-file/).

**Comment by photopea** [comment]
@syvaidya  Our JPG decoder comes from the PDF.js library. Actually, when such JPG is located inside a PDF file, Chrome PDF viewer renders it correctly, but Adobe Reader and PDF.js render it incorrectly. So I am afraid they will not fix it, as that is how Adobe software handles such JPG.

Also, if you open that raw JPG in Photoshop, it is also displayed incorrectly.

**Comment by skibulk** [comment]
I see you posted an issue in the pdf.js repo. Thank you. For reference and tracking purposes, that issue is located [here](https://github.com/mozilla/pdf.js/issues/11931).

**Comment by syvaidya** [comment]
@photopea @skibulk Looks like the issue got fixed in pdf.js. Not sure though which release it will be part of. Once it is released, we can update the library in UTIF.js too.

**Comment by photopea** [comment]
@syvaidya @skibulk I updated UTIF.js with the latest JPG decoder from PDF.js . It fixed the parsing of the two TIFFs  from the beginning of this issue.

However, it still handles the [this multipage TIFF file](https://www.nightprogrammer.org/development/multipage-tiff-example-download-test-image-file/) incorrectly (not as RGB) - white turns into pink. I extracted the JPG image and JPG viewers also display it as pink.

I think in this case, the information, that the JPG contains RGB, is not stored in a RGB file itself, but in a TIFF parameters . @syvaidya could you take a look at it? Are we constructing the JPG correctly? (copying JPG tables - t347).

**Comment by syvaidya** [comment]
@photopea I have done some analysis. Problem is that actual JPEG data in the file (inside SOS 0xffda marker) the components are not marked as RGB, but photometric interpolation in TIFF is RGB. I checked TwelveMonkeys java library to see how it handles this, and in there it will not do YCC to RGB conversion if photometric interpolation is 2 (RGB). But in our case, we don't have any way to tell jpg.js library that even though SOS marker is saying that is not RGB, actually it is RGB, and should not be doing color conversion. I think I'll have to somehow manipulate SOS data if photometric interpolation is 2. Give me some time.

**Comment by photopea** [comment]
@syvaidya  I can rewrite the JPEG decoder, to make it not convert YCC to RGB. But how do we decide when it is needed?

From what I see, the PhotometricInterpretation is always 2 for JPG inside TIFF, even for YCC data.

**Comment by syvaidya** [comment]
@photopea I was also thinking of patching JpegDecoder, but you are right. I am also not sure how we'll determine whether YCC to RGB conversion is required or not. Can you give me various TIFF samples that you have, so that I can check what other libraries are doing for each and maybe it will give some idea.

**Comment by syvaidya** [comment]
@photopea by the way, will you be okay if I convert the source code to a proper npm project, and then keep jpg.js and pako modules external to actual code? We can keep the built UTIF.js checked-in to github, but every time source changes, we'll have to do npm build.

**Comment by photopea** [comment]
@syvaidya I don't have any JPG7-TIFFs other than these three. Maybe you can find more online.

I am not a user of NPM. I really doubt the latest JPG decoder is available on npm (I copy-pasted it and manually edited some parts for better minification). The current UTIF.js does not have a JPG dependency. I can also replace the pako.js dependency with my own inflator 4 kB, 2x faster than pako.js

**Comment by syvaidya** [comment]
@photopea I understand. At least can we have all the third party code unminified within UTIF.js? It is difficult to fix something in JpegDecoder when it is minified. Maybe you can give UTIF.js which contains no minified code, and then another UTIF.min.js which has full code minified (for production use).

**Comment by jebastincharles** [comment]
@syvaidya 
Can you show some light whether the above issue is fixed?.

**Comment by syvaidya** [comment]
No it is not.
@photopea I frankly need non-minified code to be able to debug this within jpg.js. Did you think about my suggestion in last comment?

**Comment by photopea** [comment]
@syvaidya The current unminified JPEG decoder is here: https://pastebin.com/xd3kMuBB .  But I don't think we need to modify it or even know how it works.

Note, that we fixed the original problem of this issue (making UTIF.js work with the two TIFF files at the beginning). So now, we are trying to make it work only with this file: http://www.nightprogrammer.org/wp-uploads/2013/02/multipage_tiff_example.tif .

I am completely fine with closing this issue and leaving it as it is. Since the TIFF specification has never been "complete" at any point, I think we should not care about these cases.

**Comment by benwiley4000** [comment]
I'm running into this issue with a dataset of compression-7-encoded tiff images... I assume there's no news?

**Comment by photopea** [comment]
@benwiley4000 do your TIFF images work well in other programs?

**Comment by benwiley4000** [comment]
@photopea they work well in macOS Preview, which supports JPEG and LZW compression, and they were created by a data scientist who used a python image library to encode the images. Unfortunately I can't share the image set. When I converted the images to LZW they displayed great with UTIF.

LZW:

![Screen Shot 2020-11-04 at 6 39 49 PM](https://user-images.githubusercontent.com/13558253/98179626-2e5e6800-1ecd-11eb-8874-9c692f592435.png)
![Screen Shot 2020-11-04 at 6 39 44 PM](https://user-images.githubusercontent.com/13558253/98179629-30282b80-1ecd-11eb-8199-39a88a9beb92.png)

JPEG-in-TIFF:

![Screen Shot 2020-11-04 at 6 39 37 PM](https://user-images.githubusercontent.com/13558253/98179630-30c0c200-1ecd-11eb-897d-f51c64b80337.png)
![Screen Shot 2020-11-04 at 6 39 32 PM](https://user-images.githubusercontent.com/13558253/98179632-30c0c200-1ecd-11eb-93ec-2ec657f86cf3.png)

**Comment by photopea** [comment]
I think the error is the same as in our previous images.

BTW. if you can choose to what format save, why don't you save directly into JPG? The JPG inside TIFF just makes the output files bigger.

You can have multiple images in one TIFF file, but having multiple JPG files in a folder is pretty much the same.

**Comment by benwiley4000** [comment]
@photopea good question, I asked the same question to the guy building the dataset. He said he's using the TIFF basically as a container format for storing files and metadata. Ultimately I'm not the one building the dataset so I'm trying to support what users are throwing my way. 😄  But I definitely mentioned that LZW-compression has better support.

**Comment by aashay8** [comment]
+1

**Comment by thibaudDavid68** [comment]
I had the same pb, and I manage to fix it by passing value to the consructor of the JpegDecoder (in the decodeNewJPEG)

`
var param = null;   
if(img["t259"][0] === 7 && img["t262"][0] === 2 && img["t530"] && img["t530"][0] === 2 && img["t530"][1] === 2) {  
   param = {w: 0,n: 0};  
}  
var parser = new UTIF.JpegDecoder(param);  
parser.parse(buff);	 
`  
In my case the compression (t259) was 7 (new Jpeg), the PhotometricInterpretation (t262) was 2 (RGB) but the tiff aslo has the YCbCrSubSampling (t530), for what I known this is a contradtion with the tow other params.
I am not sure about this, but I home it will be usefull for other.

If it is a "goog solution" someone could explain what the use of the the parameter "w and n" ?

**Comment by photopea** [comment]
I recommend (losslessly) converting all TIFF files that you have to JPG / PNG as soon as possible, before all the software which can process these files disappears. TIFF is a very messy format.

**Comment by thibaudDavid68** [comment]
I agree, but I cannot in my context, the customers send files and  I stock and show them (i must ensure that the files remain untouched, to respect multiple reglementations). Also tif is use full to avec multiple images in the same file.

**Comment by WeiLu67** [comment]
Is there a solution for this? I have a related question #139

**Comment by photopea** [comment]
Guys, it should be fixed now thanks to @mallenLF ! :)
